### PR TITLE
Add CDC E2E test case at pipeline

### PR DESCRIPTION
**Pull request by azexcy**
Relate #22500 

Changes proposed in this pull request:
  - Add MySQL, openGauss CDC E2E test case.
  - Add build ShardingSphereDataSource method at pipeline CI
  - Fix loss timestamp precision

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            + "TYPE(NAME='hash_mod',PROPERTIES('sharding-count'='4')),"
+            + "KEY_GENERATE_STRATEGY(COLUMN=order_id,TYPE(NAME='snowflake'))"
+            + ")";
+    
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
+    
+    public AbstractCDCE2EIT(final PipelineTestParameter testParam) {
+        super(testParam);
+        // if the time zone of the unit test and the time zone of the mirror do not match, CDC will get the wrong timestamp
+        TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java:66]: It might work around for E2E, but I'm afraid it doesn't work for real world. Could we set timezone in CDC dumper/importer etc?
Since we don't know users' database timezone, we should supply a way to configure it for users.

**Review by sandynz — CHANGES_REQUESTED** [review]

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/task/CDCIncrementTask.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import java.sql.Date;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.ThreadLocalRandom;
+
+@RequiredArgsConstructor
+@Slf4j
+public final class CDCIncrementTask extends BaseIncrementTask {
```
[the remark this answers is not in the document]
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/task/CDCIncrementTask.java:35]: If reuse, `MySQLIncrementTask` and `PostgreSQLIncrementTask` , may can't merge OpenGaussCDCE2EIT and MySQLCDCE2EIT, because it's column is different now.

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/MySQLCDCE2EIT.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        startIncrementTask(new CDCIncrementTask(getProxyDataSource(), getSourceTableOrderName(), 0, 20, 3));
+        getIncreaseTaskThread().join(20000);
+        List<Map<String, Object>> actualProxyList;
+        try (Connection connection = getProxyDataSource().getConnection()) {
+            ResultSet resultSet = connection.createStatement().executeQuery(String.format("SELECT * FROM %s ORDER BY order_id ASC", getOrderTableName()));
+            actualProxyList = transformResultSetToList(resultSet);
+        }
+        // TODO wait CDC sync data finished
+        ThreadUtil.sleep(4, TimeUnit.SECONDS);
+        Awaitility.await().atMost(30, TimeUnit.SECONDS).pollInterval(2, TimeUnit.SECONDS).until(() -> listOrderRecords(getSourceTableOrderName()).size() == actualProxyList.size());
```
[the remark this answers is not in the document]
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/MySQLCDCE2EIT.java:101]: Can't reuse now  after add  query DistSQL, it will be refactor

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/framework/container/compose/DockerContainerComposer.java [hunk 4]
```diff
@@ -92,6 +92,11 @@ public String getProxyJdbcUrl(final String databaseName) {
         return DataSourceEnvironment.getURL(databaseType, proxyContainer.getHost(), proxyContainer.getFirstMappedPort(), databaseName);
     }
     
+    @Override
```
[the remark this answers is not in the document]
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/framework/container/compose/DockerContainerComposer.java:95]: now, it's equivalent, but i think `proxyContainer.getMappedPort(3307)` is more clearly

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    
+    private static final String REGISTER_STORAGE_UNIT_SQL = "REGISTER STORAGE UNIT ds_0 ( URL='${ds0}', USER='${user}', PASSWORD='${password}'),"
+            + "ds_1 ( URL='${ds1}', USER='${user}', PASSWORD='${password}')";
+    
+    private static final String CREATE_SHARDING_RULE_SQL = "CREATE SHARDING TABLE RULE t_order("
+            + "STORAGE_UNITS(ds_0,ds_1),"
+            + "SHARDING_COLUMN=user_id,"
+            + "TYPE(NAME='hash_mod',PROPERTIES('sharding-count'='4')),"
+            + "KEY_GENERATE_STRATEGY(COLUMN=order_id,TYPE(NAME='snowflake'))"
+            + ")";
```
[the remark this answers is not in the document]
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java:59]: Now there is a little difference  with migraiton DistSQL

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java (hunk 1 above)
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java:66]: Added more detailed comments

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sandynz) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/AbstractCDCE2EIT.java]: OK

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java [hunk 6]
```diff
@@ -383,4 +377,19 @@ protected void assertGreaterThanOrderTableInitRows(final int tableInitRows, fina
...
         assertTrue("actual count " + recordsCount, recordsCount > tableInitRows);
     }
+    
+    protected DataSource generateShardingSphereDataSourceFromProxy() throws SQLException {
+        String dataSourceConfigText = queryForListWithLog("EXPORT DATABASE CONFIGURATION").get(0).get("result").toString();
+        if (PipelineEnvTypeEnum.DOCKER == ENV.getItEnvType()) {
+            DockerStorageContainer storageContainer = ((DockerContainerComposer) containerComposer).getStorageContainers().get(0);
+            String sourceUrl = String.join(":", storageContainer.getNetworkAliases().get(0), Integer.toString(storageContainer.getExposedPort()));
+            String targetUrl = String.join(":", storageContainer.getHost(), Integer.toString(storageContainer.getMappedPort()));
+            dataSourceConfigText = dataSourceConfigText.replace(sourceUrl, targetUrl);
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java:387]: Could we replace jdbc url in `YamlRootConfiguration` object's data source definition, but not whole dataSourceConfigText?

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java [hunk 7]
```diff
@@ -383,4 +377,19 @@ protected void assertGreaterThanOrderTableInitRows(final int tableInitRows, fina
...
+        }
+        YamlRootConfiguration rootConfig = YamlEngine.unmarshal(dataSourceConfigText, YamlRootConfiguration.class);
+        for (Map<String, Object> each : rootConfig.getDataSources().values()) {
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java:390]: `each` should be `entry`, refer to Code of Conduct for more details

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java [hunk 8]
```diff
@@ -383,4 +377,19 @@ protected void assertGreaterThanOrderTableInitRows(final int tableInitRows, fina
         int recordsCount = getTargetTableRecordsCount(tableName);
         assertTrue("actual count " + recordsCount, recordsCount > tableInitRows);
     }
+    
+    protected DataSource generateShardingSphereDataSourceFromProxy() throws SQLException {
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java:381]: TODO could be added for `generateShardingSphereDataSourceFromProxy` method, about why it's needed.
And then it might be simplified later.

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+                getDatabaseType());
+        PipelineDataSourceWrapper sourceDataSource = new PipelineDataSourceWrapper(generateShardingSphereDataSourceFromProxy(), getDatabaseType());
+        StandardPipelineTableMetaDataLoader metaDataLoader = new StandardPipelineTableMetaDataLoader(targetDataSource);
+        PipelineTableMetaData tableMetaData = metaDataLoader.getTableMetaData(PipelineBaseE2EIT.SCHEMA_NAME, "t_order");
+        PipelineColumnMetaData primaryKeyMetaData = tableMetaData.getColumnMetaData(tableMetaData.getPrimaryKeyColumns().get(0));
+        ConsistencyCheckJobItemProgressContext progressContext = new ConsistencyCheckJobItemProgressContext("", 0);
+        SingleTableInventoryDataConsistencyChecker checker = new SingleTableInventoryDataConsistencyChecker("", sourceDataSource, targetDataSource, schemaTableName, schemaTableName,
+                primaryKeyMetaData, metaDataLoader, null, progressContext);
+        DataConsistencyCheckResult checkResult = checker.check(new DataMatchDataConsistencyCalculateAlgorithm());
+        System.out.println(checkResult);
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:166]: Could not use `println` to verify result

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+    }
+    
+    private void registerStorageUnit() throws SQLException {
+        String registerStorageUnitTemplate = REGISTER_STORAGE_UNIT_SQL.replace("${user}", getUsername())
+                .replace("${password}", getPassword())
+                .replace("${ds0}", appendExtraParam(getActualJdbcUrlTemplate(DS_0, true)))
+                .replace("${ds1}", appendExtraParam(getActualJdbcUrlTemplate(DS_1, true)));
+        addResource(registerStorageUnitTemplate);
+    }
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:175]: Could we extract single REGISTER STORATGE UNIT in `PipelineBaseE2EIT`? So we could invoke it any times in sub classes, but not hard-code it

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        ImportDataSourceParameter importDataSourceParam = new ImportDataSourceParameter(appendExtraParam(getActualJdbcUrlTemplate(DS_4, false, 0)), getUsername(), getPassword());
+        StartCDCClientParameter parameter = new StartCDCClientParameter(importDataSourceParam);
+        parameter.setAddress("localhost");
+        parameter.setPort(getContainerComposer().getProxyCDCPort());
+        parameter.setUsername(ProxyContainerConstants.USERNAME);
+        parameter.setPassword(ProxyContainerConstants.PASSWORD);
+        parameter.setDatabase("sharding_db");
+        // TODO add full=false test case later
+        parameter.setFull(true);
+        String schema = "";
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:202]: Is empty default schema required? Could it be null?

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        if (getDatabaseType().isSchemaAvailable()) {
+            schema = "test";
+        }
+        parameter.setSchemaTables(Collections.singletonList(SchemaTable.newBuilder().setTable(getSourceTableOrderName()).setSchema(schema).build()));
+        parameter.setDatabaseType(getDatabaseType().getType());
+        CompletableFuture.runAsync(() -> new CDCClient(parameter).start(), executor).whenComplete((unused, throwable) -> {
+            if (null != throwable) {
+                log.error("cdc client sync failed, ", throwable);
+            }
+            throw new RuntimeException(throwable);
```
Comment (sandynz) [inline test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:212]: Which thread will be broken after this runtime exception thrown? Is it required?

**Review by sandynz — CHANGES_REQUESTED** [review]

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java (hunk 11 above)
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:202]: Required, protobuf Message not allow to use `null`

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java (hunk 12 above)
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:212]: Removed

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java (hunk 10 above)
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java:175]: OK

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java (hunk 7 above)
Comment (azexcy) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java:390]: It's `values`, not `entrySet()`, do we need use `entry`?

Comment (sandynz) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/base/PipelineBaseE2EIT.java:403]: OK, I had a mistake

File: test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java (hunk 11 above)
Comment (sandynz) [reply test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/cdc/CDCE2EIT.java]: OK

**Review by sandynz — APPROVED** [review]
